src/test/shell/bazel/starlark_repository_test.sh

```diff
@@ -3183,4 +3183,97 @@ EOF
   bazel build @r >& $TEST_log || fail "expected bazel to succeed"
 }
 
+# Regression tests for https://github.com/bazelbuild/bazel/issues/21823.
+function test_repository_cache_concurrency() {
+  sha=cd55a062e763b9349921f0f5db8c3933288dc8ba4f76dd9416aac68acee3cb94
+
+  cat > MODULE.bazel <<'EOF'
+http_file = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
+[
+  http_file(
+    name = "repo" + str(i),
+    url = "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.5.0/bazel-skylib-1.5.0.tar.gz",
+  )
+  for i in range(100)
+]
+EOF
+  cat > BUILD <<'EOF'
+FILES = ["@repo{}//file".format(i) for i in range(100)]
+filegroup(
+  name = "files",
+  srcs = FILES,
+)
+genrule(
+  name = "unique_hashes",
+  srcs = [":files"],
+  outs = ["unique_hashes"],
+  cmd = """
+# Get the unique sha256 hashes of the files.
+sha256sum $(execpaths :files) |
+  cut -d' ' -f1 |
+  sort |
+  uniq > $@
+""".format(),
+)
+EOF
+
+  repo_cache_dir=$TEST_TMPDIR/repository_cache
+  trap 'rm -rf ${repo_cache_dir}' EXIT
+  bazel build --repository_cache="$repo_cache_dir" \
+    //:unique_hashes >& $TEST_log || fail "expected bazel to succeed"
+  assert_equals 1 "$(wc -l < bazel-bin/unique_hashes)"
+  assert_equals $sha "$(cat bazel-bin/unique_hashes)"
+
+  # Verify that the repository cache entry has been created.
+  cache_entry="$repo_cache_dir/content_addressable/sha256/$sha/file"
+  ls -R "$repo_cache_dir" >& $TEST_log
+  echo "$sha $cache_entry" | sha256sum --check || fail "sha256 mismatch"
+}
+
+function test_repository_cache_concurrency_with_integrity() {
```

Comment:
what's the difference from the previous test case?

Reply:
This one specifies a `sha256`. But I don't think the other one provides any additional coverage, so I removed it.